packages/serverpod_client/lib/src/file_uploader.dart

```diff
@@ -21,59 +22,60 @@ class FileUploader {
   /// Uploads a file contained by a [ByteData] object, returns true if
   /// successful.
   Future<bool> uploadByteData(ByteData byteData) async {
-    var stream = http.ByteStream.fromBytes(byteData.buffer.asUint8List());
-    return upload(stream, byteData.lengthInBytes);
+    final buffer = byteData.buffer;
+    var stream = http.ByteStream.fromBytes(buffer.asUint8List());
+    return _upload(stream, buffer.lengthInBytes);
   }
 
   /// Uploads a file from a [Stream], returns true if successful.
-  Future<bool> upload(Stream<List<int>> stream, int length) async {
+  Future<bool> upload(Stream<List<int>> stream, [int? length]) =>
```

Comment:
It looks like this method is missing E2E tests entirely, right?

And even if we add that, I think we also have to cover both upload "types". (How that whole pattern is configured / when which case will be entered is not entirely clear to me (outside of the actual upload)).

Reply:
That's right, so we're not missing another test for this method, but just the "second" upload type (as I only saw 1 I think).

And then also for the newly optional length, that seems like an interesting code-path to exercise (or not support for now).

As I said above, the bigger benefit is probably support uploads from `File` instead of taking `Stream` and `length` (or bytes) here, that are very likely derived from a file in practice.

Reply:
Problem with `File` is that it belongs to `dart:io` which I suppose we try to avoid in `serverpod_client` to allow clients to run on web.

Reply:
I'll add more tests.

Reply:
Ah, then maybe https://pub.dev/packages/cross_file would have to be used in practice?

In the end the important part about the API parameter would just be a “path” so that the later upload can use the most efficient way of reading/streaming it it, and we don’t have to fully read it into memory like ByteData.

I was just thinking that this is probably the 99% case, and thus it would be good to handle it once instead of burdening all callers with it.

But either way, that’s out of scope of this concrete improvement. As we have surprisingly little usage of this in this repo, I don’t feel that I can make a good case for any API update here.

Reply:
Trying to wake the AI again..

> I'll add more tests, however..
> 
> Currently multipart file upload is only used for S3 and GCP, whereas we use binary upload for the Postgres backend. Adding extra tests in this area will require some integration work towards these services. 

@coderabbitai could you make a separate issue for this?

Reply:
@Timm We use multipart upload for both S3 and GCS. Not sure why that was chosen, given that we don't actually use multiple parts. I guess a simple S3 PutObject would do. Anyway, this is getting into #3128 territory.